src/sympc/protocol/falcon/falcon.py

```diff
@@ -404,3 +432,353 @@ def multiplication_protocol(
+
+        x_wrap = Falcon.wrap3(x1, x2, x3)
+
+        # numpy random generator generate by shape only for float types.
+        r1 = torch.empty(size=shape, dtype=torch.bool).random_(generator=gen).numpy()
+        r2 = torch.empty(size=shape, dtype=torch.bool).random_(generator=gen).numpy()
+        r3 = x_wrap ^ r1 ^ r2
+
+        x_wrap_sh: List[np.ndarray] = [r1, r2, r3]
+        share_ptrs = ReplicatedSharedTensor.distribute_shares(x_wrap_sh, session, 2)
+
+        alpha = MPCTensor(shares=share_ptrs, session=session, shape=shape)
+
+        a.to_numpy(dtype)
+
+        return a, x, x_p, alpha
+
+    @staticmethod
+    def wrap(a: MPCTensor) -> MPCTensor:
+        """Falcon Wrap functionality which computes wrap on underlying shares.
+
+        Args:
+            a (MPCTensor) : input tensor to compute wrap.
+
+        Returns:
+            wrap_sh (MPCTensor): bit shares of wrap of the input tensor.
+        """
+        session = a.session
+
+        a, x, x_p, alpha = Falcon.wrap_preprocess(a, session)
+
+        r = x + a
+
+        # TODO : change to get shares by reconstruct,malicious returns all_shares
+        r1 = r.share_ptrs[0].get_copy().shares[0]
+        r2, r3 = r.share_ptrs[1].get_copy().shares
+
+        share_ptrs = []
+        for a_sh, x_sh in zip(a.share_ptrs, x.share_ptrs):
+            share_ptrs.append(a_sh.wrap_rst(x_sh))
+
+        beta = MPCTensor(shares=share_ptrs, session=session, shape=a.shape)
+
+        delta = Falcon.wrap3(r1, r2, r3)
+
+        r_public = r1 + r2 + r3
+
+        eta = Falcon.private_compare(x_p, r_public + 1)
+
+        wrap_sh = beta + delta - eta - alpha
+
+        wrap_sh.from_numpy()
+
+        return wrap_sh
+
+    @staticmethod
+    def drelu(a: MPCTensor) -> MPCTensor:
+        """Computes Derivative of ReLU on input MPCTensor.
+
+        Args:
+            a (MPCTensor): input tensor.
+
+        Returns:
+            result (MPCTensor): DReLU of input tensor.
+
+        Raises:
+            ValueError: If the input tensor does not have a valid shape.
+        """
+        session = a.session
+        shape = a.shape
+        if shape is None:
+            raise ValueError("Input MPCTensor must have valid shape.")
+
+        ring_size = session.ring_size
+        ring_bits = get_nr_bits(ring_size)
+        msb_idx = ring_bits - 1  # index of msb
+
+        share_ptrs_msb: List[ReplicatedSharedTensor] = []
+        share_ptrs_wrap: List[ReplicatedSharedTensor] = []
+
+        for share in a.share_ptrs:
+            share_ptrs_msb.append(share.bit_extraction(msb_idx))
+            share_ptrs_wrap.append(share << 1)
+
+        msb = MPCTensor(shares=share_ptrs_msb, session=session, shape=shape)
+        wrap_lshift = MPCTensor(shares=share_ptrs_wrap, session=session, shape=shape)
+
+        wrap = Falcon.wrap(wrap_lshift)
+
+        result = msb ^ wrap ^ 1
+
+        return result
+
+    @staticmethod
+    def relu(a: MPCTensor) -> MPCTensor:
```

Comment:
So the thing is that in most cases relu won't receive an MPCTensor with binary shares since relu is applied after conv2d or matmul. So lets check ring size and perform share conversion here.

Reply:
I think we do not need Hrishi it, as binary share conversion is done implictly

Reply:
The computation is done integers itself.

Reply:
https://github.com/OpenMined/SyMPC/blob/d1cb9d980db0be310c11f0eeecf83aa16e8ddc3f/tests/sympc/protocol/falcon/falcon_test.py#L350

Reply:
I forgot to include floats, I will include it,